Use augur subsample

victorlin opened the pull request:
## Description of proposed changes

The augur subsample command is built to replace the pattern of multiple intermediate augur filter calls with simpler Snakemake rules and configuration.

Note that this is a breaking change and the old configuration will no longer work. I've updated the config key from "subsampling" to "subsample" to make the breaking change more obvious, and also to keep consistent with a pattern of matching key names with rule names.

The implementation tightly couples the subsample rule with the dump of config values which was previously only used for debugging. It's not ideal, but I think it's the best way forward especially when considering future compatibility with nextstrain run.

See also #97 for a draft of future config.

<details>
<summary>
updated workflow diagram
</summary>

<img width="1427" height="827" alt="dag" src="https://github.com/user-attachments/assets/d7550b25-30db-4525-ab5e-d24af14ae487" />

</details>

## Related issue(s)

Closes #107

## Discussion threads

- [x] [How to handle multi-build config?](https://github.com/nextstrain/WNV/pull/96#discussion_r2244239263) Trivial for this workflow since config files are single-build.
- [x] [Where are config options documented and what is `defaults`?](https://github.com/nextstrain/WNV/pull/96#discussion_r2244251553)
- [x] [How do config key names map to `augur filter` options?](https://github.com/nextstrain/WNV/pull/96#discussion_r2244254085) See tables under [implementation details](https://docs.nextstrain.org/projects/augur/en/stable/usage/cli/subsample.html#implementation-details) in augur subsample docs.
    - [x] [When will `augur filter` still be used?](https://github.com/nextstrain/WNV/pull/96#discussion_r2246603388) Only time will tell, but I did my best to lay out the options in the [filtering and subsampling guide](https://docs.nextstrain.org/en/latest/guides/bioinformatics/filtering-and-subsampling.html).
- [x] [Should `augur filter` related options have its own config key?](https://github.com/nextstrain/WNV/pull/96#discussion_r2244269543) I don't think so

## Checklist

<!--
Make sure checks are successful at the bottom of the PR.

If applicable, add:
- any changes to existing tests
- any additional manual testing to confirm changes

Please add a note if you need help with adding tests.
-->

- [x] Merge+release `augur subsample`
- [x] Checks pass (docker passed; conda failure pending new version of conda-base)
- [x] Update changelog
- [ ] Pre-merge: rebase to apply fixups and replace "TBD" in changelog

<!-- 🙌 Thank you for contributing to Nextstrain! ✨ -->

joverlee521 commented on phylogenetic/rules/write_config.smk:
```diff
@@ -0,0 +1,19 @@
+
+rule write_config:
+    output:
+        config = "results/{build}/run_config.yaml",
```
Hmm, wouldn't this be writing out the same config per `{build}`? Can it be run just once and just output as `results/run_config.yaml`?

victorlin replied:
Good catch, removed `{build}` in [force-push](https://github.com/nextstrain/WNV/compare/3ffb4035ab96d45ad31be5fb772e4f3b527bd498..41271bb6d17a910b585638a58204b416159b0215)

jameshadfield commented on phylogenetic/defaults/config.yaml:
```diff
@@ -13,60 +13,47 @@ inputs:
+      group_by:
+        - region
+        - year
+      subsample_max_sequences: 3000
```
These configs seem to be a mix of wildcard-based ones (e.g. `builds: ['all-lineages']`, and `{build}` is throughout the snakefiles) but then the YAML is written in mpox-style "one config =  one build". In the context of the latter, this general pattern seems ok. For the former it seems there's plenty of different things we could do. Have you got any thoughts on how they'd look? (ncov is the obvious example here)

More generally, wildcard-dependent subsampling seems well-matched to a recent conversation (link?) where we talked about writing out machine-readable, fully-expanded configs. That is, the machine-readable subsampling for whatever wildcards are in play would look just like the highlighted block here: fully expanded and no placeholder wildcards. For WNV I'd think about changing how you write out that config to:

```py
rule subsample:
    input:
        config = expanded_config
        ...
        
def expanded_config(wildcards):
    # write out fully expanded config values to "results/{build}/config.yaml"
```

(Note that in the PR as written, `results/run_config.yaml` will be overwritten by each configfile you provided to snakemake.) There's a lot of edge-cases here, and someone's probably thought a lot more about this than me.

victorlin replied:
Thanks, this is an important point. The conversation on writing out configs is https://github.com/nextstrain/pathogen-repo-guide/issues/18, and we had a related discussion on https://github.com/nextstrain/WNV/commit/3fa4ac496df2a6f5b2e1b21a59e4f2682559e171#r163097082 (pasting below to merge with this thread):

> ---
>
> [@jameshadfield] How does this work for wildcards in config values? (Maybe WNV doesn't have to consider this?)
>
> ---
>
> [@victorlin] Do you mean like this [definition](https://github.com/nextstrain/ncov/blob/60d4bca0932a379bb88c493f88b7b3249edc581e/nextstrain_profiles/nextstrain-open/builds.yaml#L56) and [usage](https://github.com/nextstrain/ncov/blob/60d4bca0932a379bb88c493f88b7b3249edc581e/nextstrain_profiles/nextstrain-open/builds.yaml#L175) in ncov?
>
> 
> ```yaml
> region: Africa
> …
> exclude: "--exclude-where 'region!={region}'"
> ```
> 
> These are evaluated using [custom Python code](https://github.com/nextstrain/ncov/blob/60d4bca0932a379bb88c493f88b7b3249edc581e/workflow/snakemake_rules/main_workflow.smk#L183-L184) in the workflow:
> 
> ```py
> build = config["builds"][wildcards.build_name]
> value = value.format(**build)
> ```
>
> Since this is happening beyond the `config` variable which is created by Snakemake itself, the dumped YAML will keep the templated string as-is.
>
> ---

I think we're on the same page that `augur subsample` needs to take the fully expanded config, so this is mostly a Snakemake problem of providing that fully expanded config (I know you wanted to take a break from Snakemake but we can't hide from it here 😅).

The problem of "multi-build configuration" extends beyond subsampling to anything that requires build-specific config. There are at least 3 approaches:

1. Per-build Snakemake config files

    <details>
    <summary>
    mpox example
    </summary>

    [clade-i/config.yaml](https://github.com/nextstrain/mpox/blob/a5b3e9e9f344d7f7bfc840809112570cd597ec34/phylogenetic/defaults/clade-i/config.yaml)

    ```yaml
    treefix_root: "" # without a root we'll midpoint root which should work great for clade I
    root: "best"
    # Clock rate chosen via treetime inference on Clade-I data excluding Clade-Ib seqs (n=73)
    # TODO: update this once more public data is available.
    clock_rate: 1.465e-06
    clock_std_dev: 6.7e-07
    ```

    [hmpxv1/config.yaml](https://github.com/nextstrain/mpox/blob/a5b3e9e9f344d7f7bfc840809112570cd597ec34/phylogenetic/defaults/hmpxv1/config.yaml)

    ```yaml
    treefix_root: "MK783032"
    root:
      - MK783032
      - MK783030
    clock_rate: 5.7e-5
    clock_std_dev: 2e-5
    ```

    Usage:

    ```
    snakemake . --config <config.yaml>
    ```

    </details>

2. One global Snakemake config file with per-build config keys

    <details>
    <summary>
    measles example
    </summary>

    [defaults/config.yaml](https://github.com/nextstrain/measles/blob/e1d1541c3ac44bc73da5872a2721d4edd10f9651/phylogenetic/defaults/config.yaml)

    ```yaml
    filter:
        group_by: "country year"
        sequences_per_group: 20
        min_date: 1950
        min_length: 5000
    filter_N450:
        group_by: "country year"
        subsample_max_sequences: 3000
        min_date: 1950
        min_length: 400
    ```

    Usage:

    [rules/prepare_sequences_N450.smk](https://github.com/nextstrain/measles/blob/e1d1541c3ac44bc73da5872a2721d4edd10f9651/phylogenetic/rules/prepare_sequences_N450.smk#L26)

    ```py
    rule filter_N450:
        params:
            min_length = config['filter_N450']['min_length'],
    ```

    </details>

3. One global Snakemake config file with string templating


    <details>
    <summary>
    ncov example
    </summary>

    [nextstrain-open/builds.yaml](https://github.com/nextstrain/ncov/blob/60d4bca0932a379bb88c493f88b7b3249edc581e/nextstrain_profiles/nextstrain-open/builds.yaml)

    ```yaml
    builds:
      africa_1m:
        subsampling_scheme: nextstrain_region_1m
        region: Africa

    subsampling:
      nextstrain_region_1m:
        …
        exclude: "--exclude-where 'region!={region}'"
    ```

    Usage:

    ```py
    def _get_specific_subsampling_setting(setting, optional=False):
        value = _get_subsampling_settings(wildcards).get(setting, "")
        # …
        build = config["builds"][wildcards.build_name]
        value = value.format(**build)
    
    def _get_subsampling_settings(wildcards):
        subsampling_scheme = _get_subsampling_scheme_by_build_name(wildcards.build_name)
        # …
        subsampling_settings = config["subsampling"][subsampling_scheme]
        # …
        return subsampling_settings

    def _get_subsampling_scheme_by_build_name(build_name):
        return config["builds"].get(build_name, {}).get("subsampling_scheme", build_name)
    ```

    </details>

This PR is using approach 1. I would say that is the easiest to implement for workflow authors, but the least scalable to many builds. Approach 2 should be easy to adopt by passing the build-specific top level key as `--config-root`. We could maybe generalize approach 3 with something in nextstrain/shared, but I'd have to think more on that.

jameshadfield replied:
My general (not prototyped!) thinking here is that if we make each workflow write out a fully expanded config for each wildcard combination [1] then `augur subsample` can just look at that. For (1) this is easy (as shown here), for (2) it only requires a little python code and (3) is more involved. We don't have any standardisation for configs in forms (2,3) so forcing the _workflow_ to do this seems the appropriate way to go. 


[1] It's not this simple. Some workflows have rules that only use a subset of available wildcard keys, e.g. avian-flu has wildcards for subtype, segment & time but [rules](https://github.com/nextstrain/avian-flu/blob/f1657094ebf9da39660a7f4ba6a5ddf51b6a9c3e/rules/main.smk#L179-L212) which only use subtype.

joverlee521 replied:
> More generally, wildcard-dependent subsampling seems well-matched to a recent conversation (link?) where we talked about writing out machine-readable, fully-expanded configs.

This was discussed in [issue for standardized configs](https://github.com/nextstrain/pathogen-repo-guide/issues/79). Without the extra layer of config translation _prior_ to running Snakemake, +1 for the general pattern of
1. Move all config expansion to the top of the main Snakefile before running any rules.
2. The individual rules should just access the expanded config.

victorlin replied:
Thanks for linking to https://github.com/nextstrain/pathogen-repo-guide/issues/79, I think that captures the essence of this discussion. Since it's been decided that `augur subsample` will take the fully expanded config which is written to a file in #98, I'm going to mark this as resolved.

jameshadfield commented on phylogenetic/defaults/config.yaml:
```diff
@@ -13,60 +13,47 @@ inputs:
-  force_include: --exclude-all --include defaults/include.txt
+# See augur subsample documentation a full list of configuration options.
+subsample:
+  defaults:
```
> See augur subsample documentation a full list of configuration options.

Is this available somewhere?

What's defaults? Like a mixin for each scheme?

victorlin replied:
> Is this available somewhere?

Oops, I added this too soon. No docs yet, just [schema](https://github.com/nextstrain/augur/blob/6deb7b935751a6f526b637f381f2c5a35e3191a4/augur/data/schema-subsample-config.json) in source code.

> What's defaults? Like a mixin for each scheme?

Default options that apply to all samples. Sample-specific values will override defaults. Source code: [augur/subsample.py:_merge_options](https://github.com/nextstrain/augur/blob/6deb7b935751a6f526b637f381f2c5a35e3191a4/augur/subsample.py#L141-L156)

victorlin replied:
`defaults` is now introduced in a separate commit: 55d8ad9ec8c661e97a2c36ca8697474e0c59afeb

victorlin replied:
Actually the switch to defaults doesn't make sense for these configs since they all only use a single sample, so I've reverted that commit.

jameshadfield commented on phylogenetic/defaults/lineage-1A/config.yaml:
```diff
@@ -2,9 +2,22 @@ reference: "defaults/lineage-1A/reference.gb"
+subsample:
+  defaults:
+    query: is_lab_host != 'true' & lineage == '1A'
+    query_columns:
```
To double check - scheme keys, with a `--` prefix and swapping `_` to `-` are equal to augur filter args?

victorlin replied:
That's what I have now, yes. The mapping is defined in [source code](https://github.com/nextstrain/augur/blob/6deb7b935751a6f526b637f381f2c5a35e3191a4/augur/subsample.py#L15-L72).

<details>
<summary>
src
</summary>

```py
GLOBAL_CLI_OPTIONS = {
    "metadata": "--metadata",
    "metadata_chunk_size": "--metadata-chunk-size",
    "metadata_delimiters": "--metadata-delimiters",
    "metadata_id_columns": "--metadata-id-columns",
    "sequences": "--sequences",
    "sequence_index": "--sequence-index",
    "subsample_seed": "--subsample-seed",
}
"""
Mapping of argparse namespace variable name to augur filter flag.
These are sent to both intermediate and final augur filter calls.
"""


FINAL_CLI_OPTIONS = {
    "output_metadata": "--output-metadata",
    "output_sequences": "--output-sequences",
    "skip_checks": ("--skip-checks", None),
}
"""
Mapping of argparse namespace variable name to augur filter flag.
These are sent to only the final augur filter call.
"""


SAMPLE_CONFIG = {
    "exclude": "--exclude",
    "exclude_all": ("--exclude-all", None),
    "exclude_ambiguous_dates_by": "--exclude-ambiguous-dates-by",
    "exclude_where": "--exclude-where",
    "include": "--include",
    "include_where": "--include-where",
    "min_date": "--min-date",
    "max_date": "--max-date",
    "min_length": "--min-length",
    "max_length": "--max-length",
    "non_nucleotide": ("--non-nucleotide", None),
    "query": "--query",
    "query_columns": "--query-columns",
    "group_by": "--group-by",
    "group_by_weights": "--group-by-weights",
    "probabilistic_sampling": ("--probabilistic-sampling", "--no-probabilistic-sampling"),
    "sequences_per_group": "--sequences-per-group",
    "subsample_max_sequences": "--subsample-max-sequences",
}
"""
Mapping of YAML configuration key name to augur filter flag.
These are sent to only the intermediate augur filter calls.
"""


# FIXME: document that these augur filter options are unsupported:
# --priorities
# --output-log
# --output-group-by-sizes
# --output-strains
# --empty-output-reporting
```

</details>

joverlee521 replied:
I'm slightly worried about the maintenance of this config mapping...Is the plan to always keep config schema for `augur subsample` in sync with the args for `augur filter`?

victorlin replied:
I chose to keep the names identical for initial review so it's obvious which filter options are being used. I would like to rename some such as `subsample_max_sequences` → `max_sequences` ("subsample" is redundant) and `exclude` → `exclude_id_lists` ("exclude" is ambiguous with other `exclude_*` options), but figured we can bikeshed naming later.

What are the maintenance concerns? As long as `augur subsample` is wrapping `augur filter`, I don't think there's a way to avoid the mapping (unless option names are used as-is in YAML like `--subsample-max-sequences: 100`, which seems like poor user experience).

joverlee521 replied:
I'm concerned about the multiple layers here that would need to change if we ever made changes to `filter`:
1. `filter` arg 
2. `subsample` config schema
3. `subsample` SAMPLE_CONFIG mapping
4. Snakemake configs

Maybe I'm overthinking it and it's nice that the Snakemake config is fed directly to `subsample`, which has the [config schema validation](https://github.com/nextstrain/augur/blob/6deb7b935751a6f526b637f381f2c5a35e3191a4/augur/subsample.py#L255-L261). This does make me question if we're going to maintain both commands indefinitely? Once `augur subsample` exists, are we still going to use `augur filter` or is everything just going to move to config based filtering?

victorlin replied:
> I'm concerned about the multiple layers here that would need to change if we ever made changes to `filter`

`augur subsample` is an ergonomic wrapper around multiple `augur filter` calls. I think it's analogous to `augur tree` being an ergonomic wrapper around IQ-TREE, where `--nthreads` is mapped to `--threads-max`. This mapping introduces more layers and slightly added complexity on the developer side, but is a good thing for users.

The fact that we are wrapping another Augur command is even better – it means any changes to 1, 2, and 3 would be bundled in the same Augur release.

victorlin replied:
> Once `augur subsample` exists, are we still going to use `augur filter` or is everything just going to move to config based filtering?

It's true that all `augur filter` calls can be fully converted to `augur subsample` calls, but I think there will still be cases where one would prefer `augur filter`. For example, to simply filter to data within the last year, specifying `--min-date 1Y` via command line is much easier than creating the equivalent YAML config file

```yaml
samples:
  all:
    min_date: 1Y
```

But I wouldn't be surprised if all Snakemake workflows move to using `augur subsample` since the YAML configuration provides a better experience for both workflow authors and workflow users.

victorlin replied:
I've pushed documentation in PRs which should address the concerns in this thread.

- How do config key names map to `augur filter` options? See the tables under [Implementation details](https://nextstrain--1879.org.readthedocs.build/projects/augur/en/1879/usage/cli/subsample.html#implementation-details) in the augur subsample CLI docs preview.
- When will `augur filter` still be used? Only time will tell, but I did my best to lay out the options in the [filtering and subsampling guide preview](https://nextstrain--257.org.readthedocs.build/en/257/guides/bioinformatics/filtering-and-subsampling.html).

jameshadfield commented on phylogenetic/defaults/lineage-2/config.yaml:
```diff
@@ -2,9 +2,22 @@ reference: "defaults/lineage-2/reference.gb"
+      group_by:
+        - region
+        - year
+      subsample_max_sequences: 3000
```
My preference was to next these augur filter args under another key, e.g. 

```yaml
  region:
    filter:
      group_by:
        - region
        - year
      subsample_max_sequences: 3000
```

so that we didn't end up mixing together keys which are "augur filter" keys and those which are special cased. Examples of the latter (so far) might be limited to `proximity_target`/`proximity`, and perhaps `input`. If there really are only one or two special cases then this is fine, but if we think there'll be more over time I'd lean towards nesting under "filter".

victorlin replied:
I don't see a need for a config key dedicated to `augur filter` related options. That seems like an unnecessary reflection of implementation detail and doesn't provide much value to the user.

victorlin replied:
If the concern is how will options like `proximity_target`/`proximity`/`input`/etc. fit in, that's covered by config sketches in #97.

It should be fine if more sample-specific options are placed next to the ones that map to `augur filter` options. I also suspect that we'd first want to implement any sample-specific options in `augur filter` if possible.

victorlin commented:
I've moved some changes into #98 to keep this PR scoped to just switching to `augur subsample`.

victorlin committed:
Use augur subsample

The augur subsample command is built to replace the pattern of multiple
intermediate augur filter calls with simpler Snakemake rules and
configuration.

Note that this is a breaking change and the old configuration will no
longer work. I've updated the config key from "subsampling" to
"subsample" to make the breaking change more obvious, and also to keep
consistent with a pattern of matching key names with rule names.

The implementation tightly couples the subsample rule with the dump of
config values which was previously only used for debugging. It's not
ideal, but I think it's the best way forward especially when considering
future compatibility with nextstrain run.

victorlin commented:
I'll re-open this once #100 is settled.

victorlin committed:
Consolidate config code into a single file

This improves readability of the main Snakefile, especially with more
config-related operations in mind.

phylogenetic/rules/config.smk #	deleted:
phylogenetic/rules/write_config.smk #

victorlin committed:
Write run config to the same file

The timestamped file did not allow Snakemake to intelligently handle
conditional runs of the subsample rule in the case of config changes in
the subsample block.

In the future, if we really want to store each run's config in its own
timestamped file, that could be written as a second file.

victorlin committed:
fixup! Use augur subsample

genehack commented on phylogenetic/defaults/config.yaml:
```diff
@@ -19,17 +19,19 @@ build_params:
-        --exclude defaults/exclude.txt
-        --include defaults/all-lineages/include.txt
+    subsample:
+      samples:
```
just curious, why the extra level of key?

victorlin replied:
`subsample` contains everything that is given to `augur subsample`, which [requires](https://docs.nextstrain.org/projects/augur/en/stable/usage/cli/subsample.html#configuration) `samples` and optionally `defaults` (not used in the WNV config file).

joverlee521 commented on phylogenetic/rules/subsample.smk:
```diff
@@ -22,6 +22,7 @@ rule subsample:
     input:
         sequences = input_sequences,
         metadata = input_metadata,
+        config = RUN_CONFIG,
```
Huh, I was under this impression this would not work either because the file modification date will always be updated and be newer than the timestamp of the output (based on [Ignoring timestamps docs](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#ignoring-timestamps)). However, local testing shows this does work as expected.

victorlin replied:
I think, in addition to file modification time, a [hash of input file contents](https://github.com/snakemake/snakemake/blob/063aeaffdff9032306f2ae5071408d6ac1da5f08/src/snakemake/caching/hash.py#L108-L114) is at least one other factor used by Snakemake to determine whether there is a reason to re-run a rule.

<details>
<summary>
based on local testing
</summary>

```console
$ nextstrain build phylogenetic results/all-lineages/metadata_filtered.tsv --dry-run
Using workflow specific profile profiles/default for setting default command line arguments.
Saved current run config to 'results/run_config.yaml'.
host: cf237109fa78
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).

$ echo "test: 0" >> phylogenetic/defaults/config.yaml

$ nextstrain build phylogenetic results/all-lineages/metadata_filtered.tsv --dry-run
Using workflow specific profile profiles/default for setting default command line arguments.
Saved current run config to 'results/run_config.yaml'.
host: 09c087429618
Building DAG of jobs...
Job stats:
job          count
---------  -------
subsample        1
total            1

$ sed -i '' -e '$d' phylogenetic/defaults/config.yaml

$ nextstrain build phylogenetic results/all-lineages/metadata_filtered.tsv --dry-run
Using workflow specific profile profiles/default for setting default command line arguments.
Saved current run config to 'results/run_config.yaml'.
host: 9c4c103e8240
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).
```

</details>

joverlee521 replied:
Ah, thanks for the pointer. This bit is important

> Hash input files that are not generated by other jobs (sorted by hash value).

So this works as expected because `RUN_CONFIG` is not an output of another rule, further supporting the decision to _not_ write the config as part of a Snakemake rule 😄
